src/incoming-message-handler/src/use-cases/retrieveMessageFromS3.ts

```diff
@@ -0,0 +1,56 @@
+import path from "path"
+import { hasRootElement } from "shared"
+import { clean } from "shared"
+import type { S3PutObjectEvent, PromiseResult, S3GatewayInterface } from "shared-types"
+import { isError } from "shared-types"
+import type { ReceivedMessage } from "src/entities"
+import ApplicationError from "src/errors/ApplicationError"
+import readReceivedDateFromS3ObjectKey from "src/utils/readReceivedDateFromS3ObjectKey"
+import formatMessageXml from "./formatMessageXml"
+import type { ValidateS3KeyResult } from "./validateS3Key"
+import validateS3Key from "./validateS3Key"
+
+export interface ValidationResult {
+  validationResult: ValidateS3KeyResult
+}
+
+export default async (
+  event: S3PutObjectEvent,
+  s3Gateway: S3GatewayInterface
+): PromiseResult<ReceivedMessage | ValidationResult> => {
+  const { bucketName, key } = event.detail.requestParameters
+
+  const externalId = path.basename(key).replace(".xml", "")
+
+  const validationResult = validateS3Key(key)
+  if (!validationResult.isValid) {
+    return { validationResult }
+  }
+
+  let messageXml = await s3Gateway.forBucket(bucketName).getItem(key)
+
+  if (isError(messageXml)) {
+    return new ApplicationError("Error while getting the message from S3", messageXml)
+  }
+
+  try {
```

Comment:
I think this should maybe be in its own use case since this is no longer just retrieving the message from S3 - it's transforming it

Reply:
Good shout! Will move it to a separate use case.